### ISSUE TITLE
Type narrowing issues with aliases for recursive type aliases

### ISSUE THREAD

**Issue by danemtsov** [microsoft/pyright#1062]
Now that recursive type aliases are supported, I ran into some stange behavior when a recursive type alias is itself used in an alias:

```python
# pyright: strict, reportUnusedVariable=false

from typing import Dict, List, Union


JSONArray = List["JSONType"]
JSONObject = Dict[str, "JSONType"]

JSONPrimitive = Union[str, float, int, bool, None]
JSONStructured = Union[JSONArray, JSONObject]

JSONType = Union[JSONPrimitive, JSONStructured]


# Using Union: type narrowing works fine
def f1(args: Union[JSONArray, JSONObject]):
    if isinstance(args, List):
        reveal_type(args)  # List[str | float | int | bool | JSONArray | Dict[str, JSONType] | None]
                           # (which is JSONArray)
    else:
        reveal_type(args)  # Dict[str, JSONType] (which is JSONObject)


# Using type alias checking for list:
def f2(args: JSONStructured):
    if isinstance(args, List):  # <- Error:
        # Unnecessary isinstance call; "JSONArray | Dict[str, JSONType]" is never an instance of "List[Unknown]"
        reveal_type(args)  # Never
    else:
        reveal_type(args)  # JSONArray | Dict[str, JSONType]

        dargs: JSONObject = args  # <- Error:
        # Expression of type "JSONArray | Dict[str, JSONType]" cannot be assigned to declared type "JSONObject"
        #   Type "JSONArray | Dict[str, JSONType]" cannot be assigned to type "JSONObject"
        #     Type "JSONArray" cannot be assigned to type "JSONObject"


# Using type alias checking for dict:
def f3(args: JSONStructured):
    # p = args
    if isinstance(args, Dict):  # <- No errors reported, but ...
        reveal_type(args)
        # ... reports "Dict[str, JSONType]" (which is "JSONObject") but treats
        # sometimes as "JSONObject", sometimes as "Never"
    else:
        reveal_type(args)
        # reports "JSONArray" but seems to treat as "JSONArray | JSONObject"

        largs: JSONArray = args  # <- Error:
        # Expression of type "JSONArray" cannot be assigned to declared type "JSONArray"
        #   Type "JSONArray" cannot be assigned to type "JSONArray"
```

Note the different behavior between `f2` and `f3`: While in `f2`, Pyright consistently complains about the unnecessary `isinstance` call and infers _args_ to be `Never`, no error is reported in `f3` and _args_ is sometimes narrowed to `Dict[str, JSONType]`, sometimes to `Never`. I couldn't find out what that depends on but this inconsistency is apparent in the info hovers of `reveal_type`:

![inconsistent object](https://user-images.githubusercontent.com/49734973/94281171-99239780-ff4e-11ea-9899-2ec082d57b5a.png)

![inconsistent array](https://user-images.githubusercontent.com/49734973/94281301-be180a80-ff4e-11ea-99cc-0f81f1b9bbfa.png)

The complex recursion structure of `JSONType` seems to be part of the cause since this works fine:
```python
# pyright: strict, reportUnusedVariable=false

from typing import Dict, List, Union


IntDict = Dict[str, int]
IntList = List[int]
IntStructured = Union[IntDict, IntList]


def f(args: IntStructured):
    if isinstance(args, List):
        reveal_type(args)  # List[int]
    else:
        reveal_type(args)  # Dict[str, int]
```

... unless I'm missing something else.

---

Running Pyright as part of the Pylance VS Code extension:
Id: ms-python.vscode-pylance
Version: 2020.9.6
VS Marketplace Link: https://marketplace.visualstudio.com/items?itemName=ms-python.vscode-pylance

**Comment by erictraut** [comment]
Thanks for the great bug report. All of these issues will be fixed in the next release, and I've included your code sample as a unit test to ensure that it continues to work without future regressions.

**Comment by erictraut** [comment]
This is addressed in Pyright 1.1.75, which I just published. It will also be included in the next version of Pylance.
